plugins/hls-import-actions-plugin/src/Ide/Plugin/ImportActions.hs

```diff
@@ -141,42 +144,52 @@ lensProvider _  state _ CodeLensParams {_textDocument = TextDocumentIdentifier {
-   pure $ Left $ ResponseError (InR ErrorCodes_InvalidParams) "Unexpected argument for lens resolve handler: ResolveAll" Nothing
-
--- | If there are any implicit imports, provide both one code action per import
---   to make that specific import explicit, and one code action to turn them all
---   into explicit imports.
+  where mkCommand ::  PluginId -> ImportEdit -> Command
+        mkCommand pId (ImportEdit{ieResType, ieText}) =
+          let title ExplicitImport = abbreviateImportTitle ieText
+              title RefineImport = "Refine imports to " <> T.intercalate ", " (T.lines ieText)
+          in mkLspCommand pId importCommandId (title ieResType) (Just [A.toJSON rd])
+lensResolveProvider _ _ _ _ _ rd = do
+   pure $ Left $ ResponseError (InR ErrorCodes_InvalidParams) (T.pack $ "Unexpected argument for lens resolve handler: " <> show rd) Nothing
+
+-- |For explicit imports: If there are any implicit imports, provide both one
+-- code action per import to make that specific import explicit, and one code
+-- action to turn them all into explicit imports. For refine imports: If there
+-- are any reexported imports, provide both one code action per import to refine
+-- that specific import, and one code action to refine all imports.
 codeActionProvider :: Recorder (WithPriority Log) -> PluginMethodHandler IdeState 'Method_TextDocumentCodeAction
 codeActionProvider _ ideState _pId (CodeActionParams _ _ TextDocumentIdentifier {_uri} range _context)
   = pluginResponse $ do
     nfp <- getNormalizedFilePath _uri
-    (MinimalImportsResult{forCodeActions}) <-
-      handleMaybeM "Unable to run Minimal Imports"
+    (ImportActionsResult{forCodeActions}, pm) <-
+      handleMaybeM "Unable to run Import Actions"
         $ liftIO
-        $ runAction "MinimalImports" ideState $ use MinimalImports nfp
-    let relevantCodeActions = filterByRange range forCodeActions
+        $ runAction "ImportActions" ideState $ useWithStale ImportActions nfp
+    newRange <- handleMaybe "Unable to get new range"
+        $ toCurrentRange pm range
+    let relevantCodeActions = filterByRange newRange forCodeActions
         allExplicit =
-          [InR $ mkCodeAction "Make all imports explicit" (Just $ A.toJSON $ ResolveAll _uri)
+          [InR $ mkCodeAction "Make all imports explicit" (Just $ A.toJSON $ ExplicitAll _uri)
+          | not $ null relevantCodeActions ]
+        allRefine =
+          [InR $ mkCodeAction "Refine all imports" (Just $ A.toJSON $ RefineAll _uri)
           | not $ null relevantCodeActions ]
-        toCodeAction uri (_, int) =
+        toCodeAction uri (ImportAction _ int ExplicitImport) =
           mkCodeAction "Make this import explicit" (Just $ A.toJSON $ ResolveOne uri int)
-    pure $ InL ((InR . toCodeAction _uri <$> relevantCodeActions) <> allExplicit)
+        toCodeAction uri (ImportAction _  int RefineImport) =
+          mkCodeAction "Refine this import" (Just $ A.toJSON $ ResolveOne uri int)
```

Comment:
This seems wrong, it's also producing ResolveOne, shouldn't it be RefineOne or sometihng?